doc/ballast.1
.TH "ballast" "1" "21 Sept 2020" "" ""
./"################################################################
.SH "NAME"
./"################################################################
\fBballast\fP \(em a lightweight SSH load balancer with user-specific
policies
./"################################################################
.SH "SYNOPSIS"
./"################################################################
.nf
\fBballast\fP [-c FILE] ALIAS [ARGUMENT...]
.fi
.PP
.RS
Obtain the least-loaded host associated with a configured ALIAS (and
any optional ARGUMENTs it takes) from the Ballast server and connect to
it using the configured relay.  A configuration FILE other than the
system default may be specified, if desired.
.RE
.PP
.nf
\fBballast\fP [-c FILE] -l ALIAS [ARGUMENT...]
.fi
.PP
.RS
List the obtained host without connecting to it.
.RE
./"################################################################
.SH "DESCRIPTION"
./"################################################################
Ballast is a framework that was specifically designed for SSH load
balancing by taking advantage of the aliasing and proxying capabilities
available in most SSH clients.  Within a traditional SSH bastion
architecture, where internal resources can only be accessed from the
outside via a hardened bastion system, Ballast is completely transparent
to users and does not require any external modifications.  Bastions and
internal hosts must have the Ballast client installed, which can be
easily managed across internal hosts by configuration management tools.
.PP
Ballast consists of a Ballast agent on each balanced host, a Ballast
server on one or more data servers, and a Ballast client on each bastion
and internal host from which the user might access balanced hosts.  The
Ballast agent periodically collects system load information and sends it
to the Ballast server.  The server aggregates the load data received
from all agents and stores it in a suitable form for making balancing
decisions.  When the user invokes SSH to a host alias designated for
balancing, SSH triggers the Ballast client, which contacts the server to
resolve balancing aliases to actual balanced host names.  The server
consults its data and returns one or more hosts, one of which the client
connects to via a built-in TCP relay (or external equivalent such as
netcat), after which the login proceeds normally.
.PP
Unlike traditional load balancers, Ballast has the ability to adjust
its balancing strategy based on the invoking user.  By tailoring system
selections specifically to each individual user, the utility to each
user can be maximized instead of choosing a system that may be best
according to a common metric, but not ideal for the user's needs.
.PP
Prior to running the client, consult ballastrc(5) and configure client
items as needed.  The main items of interest are "alias_domain", which
specifies the domain for which aliases are relevant, "data_host", which
specifies the server host, and "hosts_X", which specifies the default
set of hosts for each alias when the server cannot be contacted.  The C
client also requires the "relay_path" setting, which is the path to a
TCP relay.
./"################################################################
.SH "SSH BALANCING"
./"################################################################
The Ballast client is intended for use on bastion and internal hosts.
Users need only know the defined balancing aliases.  To connect via ssh
to a given alias, the user simply invokes ssh to that alias.  For
example:
.PP
.RS
.nf
ssh foo
.fi
.RE
.PP
would connect to one of the hosts associated with alias "foo".  To
connect to the last host the user connected to via the balancer, the
user can append "-last" to the alias name.  For example:
.PP
.RS
.nf
ssh foo-last
.fi
.RE
.PP
would connect to the last host returned to the user for alias "foo".
Note that this suffix is configurable (see "alias_last" in
ballastrc(5)).
./"================================================================
.SS "SSH Configuration For Internal Hosts"
./"================================================================
To invoke ballast transparently, each client system must have an entry
for each balancing alias in either the global /etc/ssh/ssh_config
applicable to all users or in the per-user ~/.ssh/config.  The following
shows the relevant configuration for an alias "foo" in domain
"example.com":
.PP
.RS
.nf
# Aliases for least loaded and last utilized system
Host foo foo-last foo.example.com foo-last.example.com

    # Use the balancer when connecting to these aliases
    ProxyCommand /usr/local/bin/ballast %h

    # Avoids host key checks for all variants when
    # StrictHostKeyChecking is enabled (optional)
    HostKeyAlias foo1.example.com
.fi
.RE
.PP
A similar configuration must be defined for each alias in use (see
"hosts_X" in ballastrc(5)).  Note that "example.com" must match the
value of the configured domain (see "alias_domain" in ballastrc(5)) and
"-last" must match the configured last chosen host suffix if it differs
from the default (see "alias_last" in ballastrc(5)).  Also note that
if additional arguments are used in a balancing policy (see "arg0,
arg1, ... argN" in ballastrc(5)), the desired arguments can be placed after
the "%h" in the ProxyCommand setting.
./"================================================================
.SS "SSH Configuration For External Hosts Via Bastion"
./"================================================================
While the client can be used directly from beyond an organization's
perimeter (assuming the server's listening port is accessible), client
requests to the server are unencrypted so are subject to modification if
an attacker has control of the network along the path from client to
server.  The danger of this scenario is a modified return value from the
server that directs the client to connect to a host under the control of
the attacker, which can then be used to harvest user passwords.  SSH
host key checks, however, should prevent the user from accessing
unexpected hosts as long as the user has a key for each alias in
~/.ssh/known_hosts (through either a previous connection or manual
entry).
.PP
External clients can still effectively connect to load-balanced hosts
using appropriate SSH configuration through a bastion host.  Assuming a
bastion "bastion.example.com" that has Ballast installed and
configured, the following can be added to a user's ~/.ssh/config to
invoke ballast remotely:
.PP
.RS
.nf
# Aliases for least loaded and last utilized system
Host foo foo-last foo.example.com foo-last.example.com

    # Use the balancer on the bastion for these aliases
    ProxyCommand ssh bastion.example.com /usr/local/bin/ballast %h

    # Avoids host key checks for all variants when
    # StrictHostKeyChecking is enabled (optional)
    HostKeyAlias foo1.example.com
.fi
.RE
.PP
This configuration can also be used globally in /etc/ssh/ssh_config if
an external system has many remote users.  Note that the Ballast client
can be used as a general-purpose replacement for TCP relays like netcat
on external systems even for systems that do not correspond to balancing
aliases.
./"################################################################
.SH "GENERAL-PURPOSE BALANCING"
./"################################################################
To list the host that would have been selected for the user, the
client can be invoked directly with a -l option and the chosen alias.
For example:
.PP
.RS
.nf
ballast -l foo
.fi
.RE
.PP
would print the host that has been selected for alias "foo" to the
terminal instead of connecting to it.  This functionality allows Ballast
to act as a general-purpose load balancing solution beyond just SSH.
This functionality is also useful for testing purposes.
.PP
Additional arguments can be given to the Ballast client that will
correspond to arg0...argN in policy definitions.  For example:
.PP
.RS
.nf
ballast -l foo bar baz
.fi
.RE
.PP
would print the host that has been selected for alias "foo" when arg0
is set to "bar" and arg1 is set to "baz" in the corresponding balancing
policy.
./"################################################################
.SH "EXIT STATUS"
./"################################################################
ballast exits with 0 on success or >0 if an error occurs.  See
ballastd(1) for troubleshooting.
./"################################################################
.SH "FILES"
./"################################################################
.TP
/etc/ballastrc
This file specifies various configuration items including alias
definitions, relay settings, and server communication details.
.TP
/etc/ssh/ssh_config and/or ~/.ssh/config
Global and per-user SSH configuration files where ballast is
invoked for load balancing.
.TP
/etc/ssh/ssh_known_hosts and/or ~/.ssh/known_hosts
Global and per-user list of known SSH host keys that may be referenced
during transparent ballast invocation.
./"################################################################
.SH "AUTHOR"
./"################################################################
ballast was written by Paul Kolano.
./"################################################################
.SH "SEE ALSO"
./"################################################################
ballast-agent(1), ballastd(1), ballastrc(5), ssh_config(5)


CHANGES
CHANGES
=======

* Ballast 1.4 (07/28/10)
    - First public release

* Ballast 1.5 (07/28/10)
    - Added change log
    - Fixed missing parentheses in policy samples
    - Term "users" is now computed during data ingestion instead of
        dynamically
    - Added logging of host policy values and stale data notification
        at log level 4
    - Fixed evaluation of abs()
    - Made minor changes to docs in various places
    - Added default conf location to ballast-logstats and fixed
        exit when conf file not available
    - Modified log file parsing in ballast-logstats to be more precise
    - Readded quotes to logging of selection at log level 2
    - Fixed selection when multiple hosts evaluate to same min value

* Ballast 1.6 (--/--/--)
    - Internal release

* Ballast 1.7 (08/26/11)
    - Switched database from SDBM_File to DB_File for greatly enhanced
        stability and performance (previous databases will be invalid)
    - Added ability to use different server types based on min/max forks
    - Fixed comments in ballastd
    - Removed dist target from make file
    - Added data alias setting to allow multiple aliases with same data
    - Added random wait capability to agent
    - Added ability to use line continuation operator (\) in config file
    - Agent now tries to detect fully-qualified domain name of host
    - Added ability to use extra client arguments in policies
    - Updated documentation
    - Fixed tcp_allow and tcp_deny examples in config skeleton
    - Added ability to setuid/setgid agent
    - Added timeout to agent
    - Fixed hang during timeout with multiple data servers in C client
    - Fixed overwrite of request with multiple data servers in C client
    - Fixed backup server handling in perl client when all hosts unavailable

* Ballast 1.8 (05/13/13)
    - Fixed evaluation of user_uid (bug report by V. Gouveia)
    - Added timeout to server

* Ballast 2.0 (07/12/16)
    - Added module dependencies via App::FatPacker so can run with stock perl
    - Added built-in XXTEA encryption for communication between agent and server
    - Moved ballast-dumpdb functionality into ballastd accessible via --dump
    - Moved ballast-logstats functioality into ballastd accessible via --stats
    - Changed stats output to use Text::FormatTable instead of Text::Table
    - Fixed /proc/net/dev parsing (interface may/may not have space after)

* Ballast 2.1 (05/02/18)
    - Added abort on read/write errors to built-in relay
    - Added relay_buffer config option to control i/o size of built-in relay

* Ballast 2.2 (02/25/19)
    - Added --jsondump to dump stats in JSON format (contribution by B. Roose)
    - Added systemd service config for ballastd (contribution by B. Roose)
    - Added use_ip_address config option to reference hosts by IP
        address instead of host name (contribution by B. Roose)
    - Added user_gid variable for use in balancing policies (github issue #4)
    - Changed stats generation for better performance with large log files

* Ballast 2.3 (10/05/20)
    - Added Debian packaging files (contribution by J. Fernyhough)
    - Added installation via make
    - Changed documentation to man page format and expanded information
    - Fixed collection of user I/O info when agent becomes non-root user
    - Fixed compiler warning in C client
    - Removed makeenv.sh as it was only for AIX systems with C client


c/ballast.c
//
// Copyright (C) 2010-2018 United States Government as represented by the
// Administrator of the National Aeronautics and Space Administration
// (NASA).  All Rights Reserved.
//
// This software is distributed under the NASA Open Source Agreement
// (NOSA), version 1.3.  The NOSA has been approved by the Open Source
// Initiative.  See http://www.opensource.org/licenses/nasa1.3.php
// for the complete NOSA document.
//
// THE SUBJECT SOFTWARE IS PROVIDED "AS IS" WITHOUT ANY WARRANTY OF ANY
// KIND, EITHER EXPRESSED, IMPLIED, OR STATUTORY, INCLUDING, BUT NOT
// LIMITED TO, ANY WARRANTY THAT THE SUBJECT SOFTWARE WILL CONFORM TO
// SPECIFICATIONS, ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR
// A PARTICULAR PURPOSE, OR FREEDOM FROM INFRINGEMENT, ANY WARRANTY THAT
// THE SUBJECT SOFTWARE WILL BE ERROR FREE, OR ANY WARRANTY THAT
// DOCUMENTATION, IF PROVIDED, WILL CONFORM TO THE SUBJECT SOFTWARE. THIS
// AGREEMENT DOES NOT, IN ANY MANNER, CONSTITUTE AN ENDORSEMENT BY
// GOVERNMENT AGENCY OR ANY PRIOR RECIPIENT OF ANY RESULTS, RESULTING
// DESIGNS, HARDWARE, SOFTWARE PRODUCTS OR ANY OTHER APPLICATIONS RESULTING
// FROM USE OF THE SUBJECT SOFTWARE.  FURTHER, GOVERNMENT AGENCY DISCLAIMS
// ALL WARRANTIES AND LIABILITIES REGARDING THIRD-PARTY SOFTWARE, IF
// PRESENT IN THE ORIGINAL SOFTWARE, AND DISTRIBUTES IT "AS IS".
//
// RECIPIENT AGREES TO WAIVE ANY AND ALL CLAIMS AGAINST THE UNITED STATES
// GOVERNMENT, ITS CONTRACTORS AND SUBCONTRACTORS, AS WELL AS ANY PRIOR
// RECIPIENT.  IF RECIPIENT'S USE OF THE SUBJECT SOFTWARE RESULTS IN ANY
// LIABILITIES, DEMANDS, DAMAGES, EXPENSES OR LOSSES ARISING FROM SUCH USE,
// INCLUDING ANY DAMAGES FROM PRODUCTS BASED ON, OR RESULTING FROM,
// RECIPIENT'S USE OF THE SUBJECT SOFTWARE, RECIPIENT SHALL INDEMNIFY AND
// HOLD HARMLESS THE UNITED STATES GOVERNMENT, ITS CONTRACTORS AND
// SUBCONTRACTORS, AS WELL AS ANY PRIOR RECIPIENT, TO THE EXTENT PERMITTED
// BY LAW.  RECIPIENT'S SOLE REMEDY FOR ANY SUCH MATTER SHALL BE THE
// IMMEDIATE, UNILATERAL TERMINATION OF THIS AGREEMENT.
//

#include <netdb.h> 
#include <setjmp.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <unistd.h>
#include <sys/socket.h>

#define VERSION "2.3"

#ifndef CONF_FILE
#define CONF_FILE "/etc/ballastrc"
#endif
#ifndef DATA_SIZE
#define DATA_SIZE 1024
#endif

// note that -std=gnu99 must be used with gcc
// note that -maix64 must be used with aix

// required config settings
char *conf_req[] = {
    "alias_domain",
    "data_host",
    "relay_path",
};

// config settings and count
char **conf = NULL;
int nconf = 0;

// calling environment for timeouts
sigjmp_buf env;

// trim and store config setting in string
void putconf(char *string) {
    char *key = strdup(string);
    if (key == NULL) return;
    // trim comments
    char *tmp = index(key, '#');
    if (tmp != NULL) tmp[0] = 0;
    // trim trailing whitespace
    for (int i = strlen(key) - 1; i >= 0; --i)
        if (key[i] == ' ' || key[i] == '\t') key[i] = 0;
        else break;
    // trim leading whitespace
    key = &key[strspn(key, " \t")];
    // truncate string at whitespace
    key = strsep(&key, " \t");
    if (key == NULL) return;
    // allocate memory and assign value
    if ((conf = realloc(conf, (nconf + 1) * sizeof(char *))))
        conf[nconf++] = key;
}

// return string value assigned to key in config
char *getconf(char *key) {
    // loop in reverse order so defaults are read last
    for (int i = nconf - 1; i >= 0; --i) {
        if (!strcmp(key, conf[i])) {
            // value stored after first string terminator
            char *val = &conf[i][strlen(conf[i]) + 1];
            // trim leading whitespace and return value
            return &val[strspn(val, " \t")];
        }
    }
    return NULL;
}

// return integer value assigned to key in config
int igetconf(char *key) {
    char *val = getconf(key);
    return strtol(val, NULL, 0);
}

// send sdata to host/port in ainfo and return response in data
int getdata(const char *sdata, char *data, struct addrinfo *ainfo) {
    if (ainfo == NULL) return -1;

    // create socket
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) return -1;

    // connect to server
    if (connect(sock, ainfo->ai_addr, ainfo->ai_addrlen) < 0) return -1;
    freeaddrinfo(ainfo);

    // send request
    int n = write(sock, sdata, strlen(sdata));
    if (n < 0) return -1;

    // read reply
    bzero(data, DATA_SIZE);
    n = read(sock, data, DATA_SIZE);
    close(sock);
    if (n < 0 || n == DATA_SIZE) return -1;
    if (n > 0 && data[n - 1] == '\n') data[n - 1] = 0;
    return 0;
}

// resolve host name and return addrinfo with data_port filled in
struct addrinfo *resolve(char *data_host) {
    if (data_host == NULL) return NULL;
    struct addrinfo *ainfo;
    struct addrinfo hints;
    bzero(&hints, sizeof(hints));
    if (strspn(data_host, "0123456789.") == strlen(data_host))
        hints.ai_flags = AI_NUMERICHOST;
    // get ip address
    if (getaddrinfo(data_host, getconf("data_port"), &hints, &ainfo))
        return NULL;
    return ainfo;
}

// alarm handler that jumps back to calling environment with invoking signal
void timeout(int sig) {
    siglongjmp(env, sig);
}

int main(int argc, char *argv[]) {
    int opt;
    int opt_list = 0;
    char *conf_file = CONF_FILE;

    // parse arguments
    while ((opt = getopt(argc, argv, "c:l")) != -1) {
        switch (opt) {
            case 'c':
                conf_file = optarg;
                break;
            case 'l':
                opt_list = 1;
                break;
        }
    }
    if (optind >= argc) {
        fprintf(stderr, "ERROR: no hostname given\n");
        exit(1);
    }

    // open config file
    char data[DATA_SIZE];
    FILE *f = fopen(conf_file, "r");
    if (f == NULL) {
        fprintf(stderr, "ERROR: config file %s does not exist\n", conf_file);
        exit(1);
    }

    // store default configuration
    putconf("alias_last -last");
    putconf("alias_text unavailable");
    putconf("data_port 4411");
    putconf("data_timeout 2");
    putconf("relay_port 22");

    // store config settings
    while (fgets(data, DATA_SIZE, f) != NULL) {
        if (strlen(data) == DATA_SIZE - 1) {
            fprintf(stderr, "ERROR: config entry \"%s\" too long\n", data);
            exit(1);
        }
        if (data[strlen(data) - 1] == '\n') data[strlen(data) - 1] = 0;
        if (data[strlen(data) - 1] == '\r') data[strlen(data) - 1] = 0;
        putconf(data);
    }
    fclose(f);

    // check required config settings
    for (int i = 0; i < sizeof(conf_req) / sizeof(char *); i++) {
        char *tmp = getconf(conf_req[i]);
        if (tmp == NULL) {
            fprintf(stderr, "ERROR: config parameter %s is not defined\n",
                conf_req[i]);
            exit(1);
        }
    }

    // copy argument
    char arg[strlen(argv[optind]) + 1];
    strcpy(arg, argv[optind]);
    // chop off domain
    char *alias_domain = getconf("alias_domain");
    char *tmp = strstr(arg, alias_domain);
    if (tmp == arg + strlen(arg) - strlen(alias_domain)) tmp[0] = 0;
    // chop off last
    int last = 0;
    char *alias_last = getconf("alias_last");
    tmp = strstr(arg, alias_last);
    if (tmp == arg + strlen(arg) - strlen(alias_last)) {
        tmp[0] = 0;
        last = 1;
    }

    // find default hosts for given alias
    char hosts_alias[strlen(arg) + 7];
    sprintf(hosts_alias, "hosts_%s", arg);
    char *defaults = getconf(hosts_alias);

    if (defaults != NULL) {
        // given host matches defined alias, so load balance
        // construct balancing request
        char sdata[DATA_SIZE];
        snprintf(sdata, DATA_SIZE, "%s%s %d:%d", arg, 
            last ? getconf("alias_last") : "", getuid(), getgid());
        // concatenate optional arguments to request
        for (int i = optind + 1; i < argc; i++) {
            if (strlen(argv[i]) + 2 > DATA_SIZE - strlen(sdata)) {
                fprintf(stderr, "ERROR: excessive argument length\n");
                exit(1);
            }
            strncat(sdata, argv[i], DATA_SIZE - strlen(sdata) - 2);
        }
        strcat(sdata, "\n");

        // retrieve least loaded hosts from data server or time out
        signal(SIGALRM, timeout);
        char *host = strdup(getconf("data_host"));
        for (host = strtok(host, " "); host; host = strtok(NULL, " ")) {
            // resolve host name outside alarm to avoid reentrancy issues
            struct addrinfo *ainfo = resolve(host);
            alarm(igetconf("data_timeout"));
            if (sigsetjmp(env, 1) != SIGALRM && !getdata(sdata, data, ainfo)) break;
            // loop to try alternate data servers on failure
            alarm(0);
        }
        alarm(0);
        if (host == NULL) {
            // copy default data on failed retrieval for strtok
            strncpy(data, defaults, DATA_SIZE - 1);
            data[DATA_SIZE - 1] = 0;
        }
        if (strspn(data, " ") == strlen(data)) {
            // data is empty so no hosts available
            fprintf(stderr, "ERROR: all hosts for alias %s are %s\n",
                arg, getconf("alias_text"));
            exit(1);
        }

        // count number of hosts in data
        host = data;
        long n = 1;
        while ((host = strchr(host + 1, ' ')) != NULL) n++;

        // pick random host from data
        srand(time(NULL));
        n = rand() % n;
        for (host = strtok(data, " "); n-- > 0; host = strtok(NULL, " "));
        if (opt_list) {
            // print chosen host
            printf("%s\n", host);
            exit(0);
        } else {
            // connect to chosen host
            execl(getconf("relay_path"), getconf("relay_path"), host,
                getconf("relay_port"), NULL);
            // does not return
        }
    }

    // given host does not match alias, so pass through untouched
    if (opt_list) {
        // print given host
        printf("%s\n", argv[optind]);
    } else {
        // connect to given host
        execl(getconf("relay_path"), getconf("relay_path"), argv[optind],
            getconf("relay_port"), NULL);
        // does not return
    }
}

